src/Interpreters/Access/InterpreterGrantQuery.cpp

```diff
@@ -366,6 +366,13 @@ BlockIO InterpreterGrantQuery::execute()
     AccessRightsElements elements_to_grant, elements_to_revoke;
     collectAccessRightsElementsToGrantOrRevoke(query, elements_to_grant, elements_to_revoke);
 
+    if (query.current_grants)
+    {
+        AccessRights new_rights(elements_to_grant);
+        new_rights.makeIntersection(*current_user_access->getAccessRights());
+        elements_to_grant = new_rights.getElements();
```

Comment:
This line will cause two problems. A list returned by `getElements()` can contain elements with grant options, elements without grant options and partial revokes too. You can't just copy that list to `elements_to_grant`. First you have to filter that list and remove elements without grant options from it.

For example, if the current grants are
```
> SHOW GRANTS

GRANT SELECT ON test.* TO A WITH GRANT OPTION"
GRANT CREATE TABLE ON test.* TO A

> GRANT CURRENT GRANTS TO B
```
Here `GRANT CURRENT GRANTS TO B` should grant `SELECT`, not `CREATE TABLE`.

The second problem is with partial revokes, and this problem can't be solved easily. For example, if the current grants are
```
> SHOW GRANTS
GRANT SELECT ON *.* TO A
REVOKE SELECT ON test.* FROM A
GRANT SELECT ON test.mytable TO A

> GRANT CURRENT GRANTS TO B
```
Here `GRANT CURRENT GRANTS TO B` should grant `SELECT` on `*.*` excluding `test.*` but including `test.mytable`.

I think we could at least detect the second issue.

Reply:
For the second problem: can't we just skip `checkGrantOption` if we are granting `CURRENT GRANTS` (which should be grantable by design)? And also, I suppose we need to grant or revoke elements in this case in their original order.